### ISSUE TITLE
Do the batch normalization functions implemented differentiate between training and evaluation?

### ISSUE THREAD

**Issue by simonfqy** [deepchem/deepchem#1092]
**Note that some questions in this post are not correctly raised. I deleted those from my post below.**
In the script `deepchem/models/tensorgraph/layers.py`, there are two classes for batch normalization, `BatchNorm` and `BatchNormalization`, which you can see from here.
 https://github.com/deepchem/deepchem/blob/6f86377d9bff4541ff50ec014743d840a3e230f4/deepchem/models/tensorgraph/layers.py#L2870 

However, I think that in the `create_tensor()` functions of both classes, the cases between training and evaluating are not differentiated. Especially in the `create_tensor` function of the `BatchNorm` class which calls `tf.layers.batch_normalization()`, the "training" flag of this `TensorFlow` function is not specified. So I suppose that it will not work entirely properly.

I tried including the flag myself. I inserted one line of code (the second line below) in `create_tensor()` function of the `BatchNorm` class:
```Python
parent_tensor = inputs[0]
self.training = tf.equal(tf.constant(1.0, dtype=tf.float32), kwargs['training']) # Newly inserted
out_tensor = tf.layers.batch_normalization(parent_tensor)
```

in which the `kwargs['training']` should be `self._training_placeholder` of TensorGraph class, which should have a value of 1.0 for training case and 0.0 for evaluation/testing case.

**Comment by miaecle** [comment]
@simonfqy Good catch! I think you could add `self._training_placeholder` to the  `in_layers` parameter, then port it into the tf batch normalization function.

**Comment by simonfqy** [comment]
@miaecle Thanks for your reply.
Currently the `self._training_placeholder` is used as input to `layer.create_tensor()` in the `build()` function of `TensorGraph` class, as in this link:
https://github.com/deepchem/deepchem/blob/6f86377d9bff4541ff50ec014743d840a3e230f4/deepchem/models/tensorgraph/tensor_graph.py#L473

Are you suggesting that I can simply change that line to 
```Python
layer.create_tensor(in_layers=self._training_placeholder)
```
and also change the line in `layers.py` to
```Python
out_tensor = tf.layers.batch_normalization(training=parent_tensor)
```
where the `parent_tensor` is `self._get_input_tensors(in_layers)[0]`, which would be `self._training_placeholder` of `TensorGraph` class.   

If it is what you mean, I can go with this. However, even if it works for me now, I recommend you have a better way of handling it, because treating `self._training_placeholder` as `parent_tensor` might not be what you initially planned and could possibly break some code.
Nevertheless, I will try implementing it later today.

**Comment by simonfqy** [comment]
@miaecle Thanks for your suggestions. This would make the code dirtier, but for now it seems to be the best solution. I will do it right away. Do you know why the new line of code I introduced in my first post, i.e., `self.training = tf.equal(tf.constant(1.0, dtype=tf.float32), kwargs['training'])` does not work?

**Comment by peastman** [comment]
The placeholder is already passed to `create_tensor()` under the name `training`.  You don't need to alter the layer's inputs.  Take a look at the Dropout class for an example of how to use it: https://github.com/deepchem/deepchem/blob/master/deepchem/models/tensorgraph/layers.py#L3422-L3429

**Comment by miaecle** [comment]
@peastman Thanks for the clarification!

**Comment by simonfqy** [comment]
I have tried fixing the problem in #1108. From my very limited experiments, the modified `BatchNorm()` works okay, but not as stable as `BatchNormalization()`. I also tried fixing the division-by-zero problem associated with it. Not quite sure whether my way of handling it is correct or not. Please take a look, thanks!
